macro/vignettes/pfSI.Rmd

```diff
@@ -0,0 +1,385 @@
+
+- is_enabled, a rule to determine from the new state whether the
+  transition could happen.
+- when, a function that reads the current state to draw a next time to fire.
+  If the distribution is incomplete, this returns infinity for not firing.
+- fire, a function that modifies the state when this fires.
+
+This kind of definition decouples the transitions from each other.
+It gets rid of questions about disabling from the queue, at the expense
+of checking, at every time step, which is enabled.
+
+We could store these as named lists.
+```{r transition_list}
+program_globals <- new.env()
+program_globals$b <- 0.2
+program_globals$c <- 3.7
+
+is_enabled <- list(
+  infect = function(state) with(state, disease == "S"),
+  recover = function(state) with(state, disease == "I")
+)
+when <- list(
+  infect = function(state) rexp(1, 1/50),
+  recover = function(state) rexp(1, 1/200)
+)
+fire <- list(
+  infect = function(state, time) {
+    b <<- 0.3
+    within(state, {disease = "I"})
+    },
+  recover = function(state, time) {
+    b <<- 0.25
+    within(state, {disease = "S"})
+    }
+)
+for (set_env_idx in 1:length(fire)) {
+  environment(fire[[set_env_idx]]) <- program_globals
+}
+```
+For this individual, simulation is a loop that checks
+every transitions enabling every time. The logic for this comes
+from papers in chemical simulation that show that the first reaction
+method is much faster than optimized direct or next reaction methods
+when there are few transitions.
+```{r simulate_individual}
+# The current time and next firing times are part of the next state of the system.
+# If a firing time is Inf, that means it isn't scheduled.
+current_time <- 0
+times <- vapply(
+  when,
+  FUN = function(x) x(individual),
+  FUN.VALUE = vector(mode = "numeric", length = 1))
+
+step_cnt <- 5
+trajectory <- vector(mode = "list", length = step_cnt)
+for (step_idx in 1:step_cnt) {
+  to_fire <- which.min(times)
+  if (is.infinite(to_fire)) break
+  current_time <- times[[to_fire]]
+  trajectory[[step_idx]] <- list(event = names(to_fire), when = current_time)
+  individual <- fire[[to_fire]](individual, current_time)
+  was_enabled <- is.finite(times)
+  newly_enabled <- vapply(
+    is_enabled,
+    FUN = function(x) x(individual),
+    FUN.VALUE = vector(mode = "logical", length = 1))
+  times[was_enabled & !newly_enabled] <- Inf
+  newly_enabled <- when[!was_enabled & newly_enabled]
+  new_times <- current_time +
+    vapply(
+      newly_enabled,
+      FUN = function(x) x(individual),
+      FUN.VALUE = vector(mode = "numeric", length = 1))
+  times[names(new_times)] <- new_times
+}
+do.call(rbind, trajectory)
+```
+```{r}
+ls.str(program_globals)
+```
+
+That's all fun. So what I propose is to use a data.table because
+it can be automatically sorted by time. Put the times of transitions into
+each table row, along with individual state. The next-to-fire
+will be the next-individual-to-fire.
+```{r table_of_individuals}
```

Comment:
Will `bites` be able to be a different size that the other columns? I don't know if data.table handles ragged arrays.

Reply:
I tested that it does work. We can put the bites in the state data table and walk through the times.